modules/mod-deep-learning/DeepLearningEffect.cpp

```diff
@@ -0,0 +1,182 @@
+   sourceTracks = CreateSourceTracks(leader, sourceLabels);
+   
+   // Initiate processing buffer, most likely shorter than
+   // the length of the selection being processed.
+   Floats buffer{ leader->GetMaxBlockSize() };
+
+   // get each of the blocks we will process
+   std::vector<BlockIndex> indices = GetBlockIndices(leader, tStart, tEnd);
+   for (BlockIndex block : indices)
+   {
+      //Get a blockSize of samples (smaller than the size of the buffer)
+      sampleCount samplePos = block.first;
+      size_t blockSize = block.second;
+      double tPos = leader->LongSamplesToTime(samplePos); 
+   
+      // get a torch tensor from the leader track
+      torch::Tensor input = BuildMultichannelTensor(leader, buffer.get(), 
+                                            samplePos, blockSize);
+
+      // if we're not doing a multichannel forward pass, downmix
+      if (!mModel->GetCard()->multichannel())
+         input = input.sum(0, true, torch::kFloat);
+
+      // resample!
+      input = mModel->Resample(input, origRate, mModel->GetSampleRate());
+
+      // forward pass!
+      torch::Tensor output = mModel->ToTensor(ForwardPassInThread(input));
+
+      // resample back
+      output = mModel->Resample(output, mModel->GetSampleRate(), origRate);
+
+      // write each source output to the source tracks
+      for (size_t idx = 0; idx < output.size(0) ; idx++)
+         TensorToTrack(output[idx].unsqueeze(0), sourceTracks[idx], 
+                       tPos, tEnd); 
+
+      // Update the Progress meter
+      mCurrentProgress = (tPos - tStart) / (tEnd - tStart);
+      if (TrackProgress(mCurrentTrackNum, mCurrentProgress)) 
+         return false;
+   }
+
+   // postprocess the source tracks to the user's sample rate and format
+   PostProcessSources(leader, sourceTracks, origFmt, origRate);
+
+   return true;
+}
+
+std::vector<WaveTrack::Holder> DeepLearningEffect::CreateSourceTracks
+(WaveTrack *leader, std::vector<std::string> &labels)
+{
+   std::vector<WaveTrack::Holder> sources;
+   for (auto &label : labels)
+   {
+      WaveTrack::Holder srcTrack = leader->EmptyCopy();
+
+      // append the source name to the track's name
+      srcTrack->SetName(srcTrack->GetName() + wxString("-" + label));
+      sources.emplace_back(srcTrack);
+   }
+   return sources;
+}
+
+void DeepLearningEffect::PostProcessSources
+(WaveTrack *leader, std::vector<WaveTrack::Holder> &sourceTracks, sampleFormat fmt, int sampleRate)
+{
+   // flush all output track buffers
+   // convert to the original rate and format
+   for (std::shared_ptr<WaveTrack> track : sourceTracks)
+   {
+      track->Flush();
+      track->ConvertToSampleFormat(fmt);
+      track->Resample(sampleRate);
+      AddToOutputTracks(track);
+
+      // if the parent track used to be stereo,
+      // make the source mono anyway
+      mOutputTracks->UnlinkChannels(*track);
+
+      // join the clips!
+      for (const auto &timestamp : GetClipTimestamps(leader, 
```

Comment:
Why was this postprocessing step important?

I don't like it that the several effect subclasses have, some of them, special clip-handling logic, rather than having a generalized logic defined in just one place.  But I'm not sure what that should be, given possible time-stretching.

I'm not telling you to change anything, just noting something that bothers me.